File: reference.net/db4o-reference-chapters/src/Db4objects.Db4odoc/Aliases/AliasExample.cs
/* Copyright (C) 2004 - 2007 db4objects Inc. http://www.db4o.com */
using System;
using System.IO;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;

namespace Db4objects.Db4odoc.Aliases
{
    class AliasExample
    {
        private static readonly string Db4oFileName = "reference.db4o";
        private static TypeAlias tAlias;

        public static void Main(string[] args)
        {
            IConfiguration configuration = ConfigureClassAlias();
            SaveDrivers(configuration);
            RemoveClassAlias(configuration);
            GetPilots(configuration);
            SavePilots(configuration);
            configuration = ConfigureAlias();
            GetObjectsWithAlias(configuration);
        }
        // end Main

        private static IConfiguration ConfigureClassAlias()
        {
            // create a new alias
            tAlias = new TypeAlias("Db4objects.Db4odoc.Aliases.Pilot, Db4objects.Db4odoc", "Db4objects.Db4odoc.Aliases.Driver, Db4objects.Db4odoc");
            // add the alias to the db4o configuration 
            IConfiguration configuration = Db4oFactory.NewConfiguration();
            configuration.AddAlias(tAlias);
            // check how does the alias resolve
            Console.WriteLine("Stored name for Db4objects.Db4odoc.Aliases.Driver: " + tAlias.ResolveRuntimeName("Db4objects.Db4odoc.Aliases.Driver, Db4objects.Db4odoc"));
            Console.WriteLine("Runtime name for Db4objects.Db4odoc.Aliases.Pilot: " + tAlias.ResolveStoredName("Db4objects.Db4odoc.Aliases.Pilot, Db4objects.Db4odoc"));
            return configuration;
        }
        // end ConfigureClassAlias


        private static void RemoveClassAlias(IConfiguration configuration)
        {
            configuration.RemoveAlias(tAlias);
        }
        // end RemoveClassAlias

        private static void SaveDrivers(IConfiguration configuration)
        {
            File.Delete(Db4oFileName);
            IObjectContainer db = Db4oFactory.OpenFile(configuration, Db4oFileName);
            try
            {
                Driver driver = new Driver("David Barrichello", 99);
                db.Set(driver);
                driver = new Driver("Kimi Raikkonen", 100);
                db.Set(driver);
            }
            finally
            {
                db.Close();
            }
        }
        // end SaveDrivers

        private static void SavePilots(IConfiguration configuration)
        {
            File.Delete(Db4oFileName);
            IObjectContainer db = Db4oFactory.OpenFile(configuration, Db4oFileName);
            try
            {
                Pilot pilot = new Pilot("David Barrichello", 99);
                db.Set(pilot);
                pilot = new Pilot("Kimi Raikkonen", 100);
                db.Set(pilot);
            }
            finally
            {
                db.Close();
            }
        }
        // end SavePilots

        private static void GetPilots(IConfiguration configuration)
        {
            IObjectContainer db = Db4oFactory.OpenFile(configuration, Db4oFileName);
            try
            {
                IObjectSet result = db.Query(typeof(Pilot));
                ListResult(result);
            }
            finally
            {
                db.Close();
            }
        }
        // end GetPilots

        private static void GetObjectsWithAlias(IConfiguration configuration)
        {
            IObjectContainer db = Db4oFactory.OpenFile(configuration, Db4oFileName);
            try
            {
                IObjectSet result = db.Query(typeof(Db4objects.Db4odoc.Aliases.NewAlias.Pilot));
                ListResult(result);
            }
            finally
            {
                db.Close();
            }
        }
        // end GetObjectsWithAlias

        private static void ListResult(IObjectSet result)
        {
            Console.WriteLine(result.Size());
            while (result.HasNext())
            {
                Console.WriteLine(result.Next());
            }
        }
        // end ListResult

        private static IConfiguration ConfigureAlias()
        {
            // Db4objects.Db4odoc.Aliases.* - namespace for the classes saved in the database
            // Db4objects.Db4odoc.Aliases.NewAlias.* - runtime namespace
            WildcardAlias wAlias = new WildcardAlias("Db4objects.Db4odoc.Aliases.*", "Db4objects.Db4odoc.Aliases.NewAlias.*");
            // add the alias to the configuration
            IConfiguration configuration = Db4oFactory.NewConfiguration();
            configuration.AddAlias(wAlias);
            Console.WriteLine("Stored name for Db4objects.Db4odoc.Aliases.NewAlias.Pilot: " + wAlias.ResolveRuntimeName("Db4objects.Db4odoc.Aliases.NewAlias.Pilot"));
            Console.WriteLine("Runtime name for Db4objects.Db4odoc.Aliases.Pilot: " + wAlias.ResolveStoredName("Db4objects.Db4odoc.Aliases.Pilot"));
            return configuration;
        }
        // end ConfigureAlias

    }
}

File: reference/src/com/db4odoc/aliases/AliasExample.java
/* Copyright (C) 2004 - 2007 db4objects Inc. http://www.db4o.com */

package com.db4odoc.aliases;

import java.io.File;

import com.db4o.Db4o;
import com.db4o.ObjectContainer;
import com.db4o.ObjectSet;
import com.db4o.config.Configuration;
import com.db4o.config.TypeAlias;
import com.db4o.config.WildcardAlias;

public class AliasExample {
	private static final String DB4O_FILE_NAME = "reference.db4o";
	private static TypeAlias tAlias;

	public static void main(String args[]) 
	{ 
		Configuration configuration = configureClassAlias();
		saveDrivers(configuration);
		removeClassAlias(configuration);
		getPilots(configuration);
		savePilots(configuration);
		configuration = configureAlias();
		getObjectsWithAlias(configuration);
	}
	// end main
	
	private static Configuration configureClassAlias(){
		// create a new alias
		tAlias = new TypeAlias("com.db4odoc.aliases.Pilot","com.db4odoc.aliases.Driver");
		// add the alias to the db4o configuration
		Configuration configuration = Db4o.newConfiguration();
		configuration.addAlias(tAlias);
		// check how does the alias resolve
		System.out.println("Stored name for com.db4odoc.aliases.Driver: "+tAlias.resolveRuntimeName("com.db4odoc.aliases.Driver"));
		System.out.println("Runtime name for com.db4odoc.aliases.Pilot: "+tAlias.resolveStoredName("com.db4odoc.aliases.Pilot"));
		return configuration;
	}
	// end configureClassAlias
	
	
	private static void removeClassAlias(Configuration configuration){
		configuration.removeAlias(tAlias);
	}
	// end removeClassAlias
	
	private static void saveDrivers(Configuration configuration){
		new File(DB4O_FILE_NAME ).delete();
		ObjectContainer container = Db4o.openFile(configuration, DB4O_FILE_NAME);
		try {
			Driver driver = new Driver("David Barrichello",99);
			container.set(driver);
			driver = new Driver("Kimi Raikkonen",100);
			container.set(driver);
		} finally {
			container.close();
		}
	}
	// end saveDrivers
	
	private static void savePilots(Configuration configuration){
		new File(DB4O_FILE_NAME ).delete();
		ObjectContainer container = Db4o.openFile(configuration, DB4O_FILE_NAME);
		try {
			Pilot pilot = new Pilot("David Barrichello",99);
			container.set(pilot);
			pilot = new Pilot("Kimi Raikkonen",100);
			container.set(pilot);
		} finally {
			container.close();
		}
	}
	// end savePilots
	
	private static void getPilots(Configuration configuration){
		ObjectContainer container = Db4o.openFile(configuration, DB4O_FILE_NAME);
		try {
			ObjectSet result = container.query(com.db4odoc.aliases.Pilot.class);
			listResult(result);
		} finally {
			container.close();
		}
	}
	// end getPilots
	
	private static void getObjectsWithAlias(Configuration configuration){
		ObjectContainer container = Db4o.openFile(configuration, DB4O_FILE_NAME);
		try {
			ObjectSet result = container.query(com.db4odoc.aliases.newalias.Pilot.class);
			listResult(result);
		} finally {
			container.close();
		}
	}
	// end getObjectsWithAlias

	private static void listResult(ObjectSet result) {
        System.out.println(result.size());
        while(result.hasNext()) {
            System.out.println(result.next());
        }
    }
    // end listResult
	
	private static Configuration configureAlias() {
    	// com.db4odoc.aliases.* - package for the classes saved in the database
    	// com.db4odoc.aliases.newalias.* - runtime package
		WildcardAlias wAlias = new WildcardAlias("com.db4odoc.aliases.*","com.db4odoc.aliases.newalias.*");
		// add the alias to the configuration
		Configuration configuration = Db4o.newConfiguration();
		configuration.addAlias(wAlias);
		System.out.println("Stored name for com.db4odoc.aliases.newalias.Pilot: "+wAlias.resolveRuntimeName("com.db4odoc.aliases.newalias.Pilot"));
		System.out.println("Runtime name for com.db4odoc.aliases.Pilot: "+wAlias.resolveStoredName("com.db4odoc.aliases.Pilot"));
		return configuration;
    }
    // end configureAlias
}
